File: KotlinLibraryX/library/src/main/java/com/lee/library/widget/nav/BottomNavView.java
package com.lee.library.widget.nav;

import android.content.Context;
import android.graphics.Canvas;
import android.os.Build;
import android.util.AttributeSet;
import android.view.Gravity;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;

import androidx.annotation.NonNull;
import androidx.viewpager.widget.ViewPager;

import com.google.android.material.bottomnavigation.BottomNavigationItemView;
import com.google.android.material.bottomnavigation.BottomNavigationMenuView;
import com.google.android.material.bottomnavigation.BottomNavigationView;
import com.google.android.material.bottomnavigation.LabelVisibilityMode;
import com.lee.library.utils.ReflexUtil;
import com.lee.library.utils.SizeUtil;

import java.util.ArrayList;
import java.util.List;

/**
 * @author jv.lee
 * @date 2019/5/7
 * @description 使用png 等多色彩图片时 需要动态设置 itemIconTintList = null
 */
@androidx.annotation.RequiresApi(api = Build.VERSION_CODES.HONEYCOMB)
public class BottomNavView extends BottomNavigationView implements BottomNavigationView.OnNavigationItemSelectedListener {

    private ViewPager mViewPager;
    private List<NumberDotView> numberDots;
    private List<DotView> dots;
    private ItemPositionListener mItemPositionListener;

    public BottomNavView(Context context) {
        this(context, null);
    }

    public BottomNavView(Context context, AttributeSet attrs) {
        this(context, attrs, 0);
    }

    public BottomNavView(Context context, AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
        setHorizontalFadingEdgeEnabled(false);
        setLabelVisibilityMode(LabelVisibilityMode.LABEL_VISIBILITY_LABELED);
        setOnNavigationItemSelectedListener(this);
        post(() -> {
            createNumberDotViews();
            createDotViews();
        });
    }

    @Override
    protected void onMeasure(int widthMeasureSpec, int heightMeasureSpec) {
        super.onMeasure(widthMeasureSpec, heightMeasureSpec);
    }

    @Override
    protected void onDraw(Canvas canvas) {
        super.onDraw(canvas);
    }

    @Override
    public boolean onNavigationItemSelected(@NonNull MenuItem menuItem) {
        for (int i = 0; i < getMenu().size(); i++) {
            if (getMenu().getItem(i) == menuItem) {
                if (mViewPager != null) {
                    mViewPager.setCurrentItem(i, false);
                }
                if (mItemPositionListener != null) {
                    mItemPositionListener.onPosition(menuItem, i);
                }
            }
        }
        return true;
    }

    public void toPosition(int position) {
        setSelectedItemId(getMenu().getItem(position).getItemId());
    }

    public void createNumberDotViews() {
        //初始化红点view
        BottomNavigationMenuView menuView = null;
        for (int i = 0; i < getChildCount(); i++) {
            View child = getChildAt(i);
            if (child instanceof BottomNavigationMenuView) {
                menuView = (BottomNavigationMenuView) child;
                break;
            }
        }
        if (menuView != null) {
            numberDots = new ArrayList<>();
            for (int i = 0; i < menuView.getChildCount(); i++) {
                BottomNavigationItemView itemView = (BottomNavigationItemView) menuView.getChildAt(i);
                LayoutParams params = new LayoutParams(ViewGroup.LayoutParams.WRAP_CONTENT, ViewGroup.LayoutParams.WRAP_CONTENT);
                ImageView imageView = ReflexUtil.INSTANCE.reflexField(itemView, "icon");
                if (imageView != null) {
                    params.topMargin = imageView.getTop();
                    params.leftMargin = imageView.getRight();
                }
                NumberDotView dotView = new NumberDotView(getContext());
                itemView.addView(dotView, params);
                numberDots.add(dotView);
            }
        }
    }

    private void createDotViews() {
        //初始化红点view
        BottomNavigationMenuView menuView = null;
        for (int i = 0; i < getChildCount(); i++) {
            View child = getChildAt(i);
            if (child instanceof BottomNavigationMenuView) {
                menuView = (BottomNavigationMenuView) child;
                break;
            }
        }
        if (menuView != null) {
            dots = new ArrayList<>();
            for (int i = 0; i < menuView.getChildCount(); i++) {
                BottomNavigationItemView itemView = (BottomNavigationItemView) menuView.getChildAt(i);
                LayoutParams params = new LayoutParams(SizeUtil.dp2px(getContext(), 6), SizeUtil.dp2px(getContext(), 6));
                params.gravity = Gravity.CENTER_HORIZONTAL;
                params.topMargin = SizeUtil.dp2px(getContext(), 6);
                params.leftMargin = SizeUtil.dp2px(getContext(), 6);
                DotView dotView = new DotView(getContext());
                dotView.setVisibility(GONE);
                itemView.addView(dotView, params);
                dots.add(dotView);
            }
        }
    }

    public void setNumberDot(final int index, final int number) {
        if (numberDots != null && numberDots.size() > index) {
            numberDots.get(index).setNumberCount(number);
        }
        postDelayed(() -> numberDots.get(index).setNumberCount(number), 100);
    }

    public void setDotVisibility(int index, int visibility) {
        if (dots != null && dots.size() > index) {
            dots.get(index).setVisibility(visibility);
        }
        postDelayed(() -> dots.get(index).setVisibility(visibility), 100);
    }

    public interface ItemPositionListener {
        void onPosition(MenuItem menuItem, int position);
    }

    public void setItemPositionListener(ItemPositionListener mItemPositionListener) {
        this.mItemPositionListener = mItemPositionListener;
    }

    public void bindViewPager(ViewPager mViewPager) {
        this.mViewPager = mViewPager;
    }
}

File: KotlinLibraryX/library/src/main/java/com/lee/library/widget/nav/BottomNavView.kt
package com.lee.library.widget.nav

import android.content.Context
import android.graphics.Color
import android.os.Build
import android.util.AttributeSet
import android.util.TypedValue
import android.view.Gravity
import android.view.MenuItem
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import androidx.annotation.RequiresApi
import androidx.core.view.children
import androidx.viewpager.widget.ViewPager
import androidx.viewpager2.widget.ViewPager2
import com.google.android.material.bottomnavigation.BottomNavigationItemView
import com.google.android.material.bottomnavigation.BottomNavigationMenuView
import com.google.android.material.bottomnavigation.BottomNavigationView
import com.google.android.material.bottomnavigation.LabelVisibilityMode
import com.lee.library.R
import com.lee.library.extensions.dp2px
import com.lee.library.utils.ReflexUtil.reflexField

/**
 * @author jv.lee
 * @date 2019/5/7
 * @description 使用png 等多色彩图片时 需要动态设置 itemIconTintList = null
 */
@RequiresApi(api = Build.VERSION_CODES.HONEYCOMB)
class BottomNavView @JvmOverloads constructor(
    context: Context,
    attrs: AttributeSet? = null,
    defStyleAttr: Int = 0
) : BottomNavigationView(context, attrs, defStyleAttr),
    BottomNavigationView.OnNavigationItemSelectedListener {
    private var mViewPager: ViewPager? = null
    private var mViewPager2: ViewPager2? = null
    private var mItemPositionListener: ItemPositionListener? = null
    private var numberDots = ArrayList<NumberDotView>()
    private var dots = ArrayList<DotView>()
    private var isNumberInit = false
    private var isDotsInit = false

    private var dotNumberTextSize: Float = 12F
    private var dotBackground: Int = Color.RED
    private var dotNumberTextColor: Int = Color.WHITE
    private var dotLineColor: Int = Color.WHITE
    private var dotSize: Int = 6

    init {
        attrs?.let {
            context.obtainStyledAttributes(it, R.styleable.BottomNavView).apply {
                dotNumberTextSize = getDimension(R.styleable.BottomNavView_dotNumberTextSize, 12f)
                dotBackground =
                    getColor(R.styleable.BottomNavView_dotBackground, Color.RED)
                dotNumberTextColor =
                    getColor(R.styleable.BottomNavView_dotNumberTextColor, Color.WHITE)
                dotLineColor = getColor(R.styleable.BottomNavView_dotLineColor, Color.WHITE)
                dotSize = getDimension(R.styleable.BottomNavView_dotSize, 6f).toInt()
                recycle()
            }
        }
        isHorizontalFadingEdgeEnabled = false
        labelVisibilityMode = LabelVisibilityMode.LABEL_VISIBILITY_LABELED
        setOnNavigationItemSelectedListener(this)
        post {
            createNumberDotViews()
            createDotViews()
        }
    }

    override fun onNavigationItemSelected(menuItem: MenuItem): Boolean {
        for (i in 0 until menu.size()) {
            if (menu.getItem(i) === menuItem) {
                mViewPager?.setCurrentItem(i, false)
                mViewPager2?.setCurrentItem(i,false)
                mItemPositionListener?.onPosition(menuItem, i)
            }
        }
        return true
    }

    fun toPosition(position: Int) {
        selectedItemId = menu.getItem(position).itemId
    }

    private fun createNumberDotViews() {
        //初始化红点view
        var menuView: BottomNavigationMenuView? = null
        for (i in 0 until childCount) {
            val child = getChildAt(i)
            if (child is BottomNavigationMenuView) {
                menuView = child
                break
            }
        }
        menuView?.run {
            numberDots.clear()
            for (i in 0 until childCount) {
                val itemView =
                    menuView.getChildAt(i) as BottomNavigationItemView
                val params = LayoutParams(
                    ViewGroup.LayoutParams.WRAP_CONTENT,
                    ViewGroup.LayoutParams.WRAP_CONTENT
                )
                val imageView =
                    reflexField<ImageView>(itemView, "icon")
                imageView?.let {
                    params.topMargin = it.top / 2
                    params.leftMargin = (it.right - (it.width * 0.25)).toInt()
                }
                val dotView = NumberDotView(context)
                dotView.setDotLineColor(dotLineColor)
                dotView.setTextColor(dotNumberTextColor)
                dotView.setDotBackgroundColor(dotBackground)
                dotView.setTextSize(TypedValue.COMPLEX_UNIT_SP, dotNumberTextSize)
                if (!itemView.children.contains(dotView)) itemView.addView(dotView, params)
                numberDots.add(dotView)
            }
            isNumberInit = true
        }
    }

    private fun createDotViews() {
        //初始化红点view
        var menuView: BottomNavigationMenuView? = null
        for (i in 0 until childCount) {
            val child = getChildAt(i)
            if (child is BottomNavigationMenuView) {
                menuView = child
                break
            }
        }
        menuView?.run {
            dots.clear()
            for (i in 0 until childCount) {
                val itemView = menuView.getChildAt(i) as BottomNavigationItemView
                val params = LayoutParams(context.dp2px(dotSize).toInt(), context.dp2px(dotSize).toInt())
                params.gravity = Gravity.CENTER_HORIZONTAL
                params.topMargin = context.dp2px(dotSize).toInt()
                params.leftMargin = context.dp2px(dotSize).toInt()
                val dotView = DotView(context)
                dotView.setDotColor(dotBackground)
                dotView.visibility = View.GONE
                if (!itemView.children.contains(dotView)) itemView.addView(dotView, params)
                dots.add(dotView)
            }
            isDotsInit = true
        }
    }

    private fun toNumberDot(index: Int, number: Int) {
        if (numberDots.size > index) numberDots[index].setNumber(number)
    }

    private fun toDotVisibility(index: Int, visibility: Int) {
        if (dots.size > index) dots[index].visibility = visibility
    }

    fun setNumberDot(index: Int, number: Int) {
        if (isNumberInit) {
            toNumberDot(index, number)
        } else {
            postDelayed({ toNumberDot(index, number) }, 100)
        }
    }

    fun setDotVisibility(index: Int, visibility: Int) {
        if (isDotsInit) {
            toDotVisibility(index, visibility)
        } else {
            postDelayed({ toDotVisibility(index, visibility) }, 100)
        }
    }

    interface ItemPositionListener {
        fun onPosition(menuItem: MenuItem?, position: Int)
    }

    fun setItemPositionListener(itemPositionListener: ItemPositionListener?) {
        this.mItemPositionListener = itemPositionListener
    }

    fun bindViewPager(viewPager: ViewPager?) {
        this.mViewPager = viewPager
    }

    fun bindViewPager(viewPager: ViewPager2?) {
        this.mViewPager2 = viewPager
    }

}